File: src/issue_orchestrator/view_models/tech_lead_activity.py
"""Projection of the local tech-lead run history onto the dashboard (#6858).

ADR-0033's second surface. :mod:`.tech_lead_run_actions` projects what the
operator can DO right now; this projects what the tech lead has already done —
the "tech-lead activity" view the ADR names, and the thing that makes a run a
visible object rather than a session on a bookkeeping issue nobody reads.

Presentation only. Nothing here decides anything: the phases come from the
recorded run, and every string this emits is rendered verbatim so the surface
is legible without a colour key (the same rule the run-action affordances
follow). The subject is shown as a REFERENCE — "#123", a link out — never as
the run's own identity, because on a client's board that issue belongs to them.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from pydantic import BaseModel, ConfigDict, Field

from ..domain.tech_lead_run_record import TechLeadRunPhase
from ..domain.tech_lead_session import TechLeadSessionFlavor

if TYPE_CHECKING:
    from ..domain.tech_lead_run_record import TechLeadRunRecord
    from ..ports.tech_lead_run_record_store import TechLeadRunHistoryReader


# How many runs the panel shows. Small on purpose: this is a "what has the tech
# lead been doing?" glance, not an audit log — the full history stays queryable
# in the local store.
ACTIVITY_LIMIT = 20

# Human names for the three run flavors. Here rather than in the browser so the
# vocabulary has one owner: a flavor renamed in the domain cannot leave the UI
# quietly showing the old word.
_FLAVOR_LABELS: dict[TechLeadSessionFlavor, str] = {
    TechLeadSessionFlavor.HEALTH_REVIEW: "Health review",
    TechLeadSessionFlavor.BATCH_REVIEW: "Batch review",
    TechLeadSessionFlavor.FAILURE_INVESTIGATION: "Failure investigation",
}

# Colour-independent phase words, per the project's rule that status must never
# be signalled by tint alone.
_PHASE_LABELS: dict[TechLeadRunPhase, str] = {
    TechLeadRunPhase.RUNNING: "Running",
    TechLeadRunPhase.COMPLETED: "Completed",
    TechLeadRunPhase.NEEDS_HUMAN: "Needs human",
    TechLeadRunPhase.FAILED: "Failed",
    TechLeadRunPhase.WITHDRAWN: "Withdrawn",
}

# The tone each phase renders with. A CLASS name, not a colour: the stylesheet
# owns the palette, and both themes get the same four semantic buckets.
_PHASE_TONES: dict[TechLeadRunPhase, str] = {
    TechLeadRunPhase.RUNNING: "active",
    TechLeadRunPhase.COMPLETED: "good",
    TechLeadRunPhase.NEEDS_HUMAN: "warn",
    TechLeadRunPhase.FAILED: "bad",
    TechLeadRunPhase.WITHDRAWN: "muted",
}


class TechLeadRunActivityEntry(BaseModel):
    """One recorded run, as the activity panel renders it."""

    model_config = ConfigDict(extra="forbid", frozen=True, populate_by_name=True)

    run_key: str = Field(serialization_alias="runKey")
    # "Health review" / "Batch review" / "Failure investigation".
    flavor_label: str = Field(serialization_alias="flavorLabel")
    phase: str
    phase_label: str = Field(serialization_alias="phaseLabel")
    # Semantic bucket for styling. Never the only signal — ``phase_label`` is
    # always rendered too.
    tone: str
    started_at: str = Field(serialization_alias="startedAt")
    # "" while the run is still going.
    ended_at: str = Field(serialization_alias="endedAt")
    # 0 when the run subject is the whole board and no anchor was involved.
    subject_issue_number: int = Field(serialization_alias="subjectIssueNumber")
    subject_title: str = Field(serialization_alias="subjectTitle")
    detail: str
    findings: int
    proposals: int
    # Drill-down identity: what the session-replay surface needs to find this
    # run's transcript. Published rather than reconstructed in the browser so
    # the UI never has to know how run artifacts are addressed.
    run_id: str = Field(serialization_alias="runId")
    session_name: str = Field(serialization_alias="sessionName")


class TechLeadActivityView(BaseModel):
    """The dashboard's tech-lead activity panel."""

    model_config = ConfigDict(extra="forbid", frozen=True, populate_by_name=True)

    entries: tuple[TechLeadRunActivityEntry, ...]
    # The sentence shown when there are none. Published rather than hardcoded in
    # the browser: "nothing has run yet" and "this engine has no history" read
    # the same to the UI but not to an operator.
    empty_message: str = Field(serialization_alias="emptyMessage")

    @classmethod
    def empty(cls) -> "TechLeadActivityView":
        return cls(entries=(), empty_message=EMPTY_MESSAGE)


EMPTY_MESSAGE = "No tech-lead runs recorded yet."


def read_tech_lead_activity(
    history: "TechLeadRunHistoryReader", *, limit: int = ACTIVITY_LIMIT
) -> TechLeadActivityView:
    """Project the local run history onto the activity panel."""
    entries = tuple(_entry(record) for record in history.recent(limit=limit))
    return TechLeadActivityView(entries=entries, empty_message=EMPTY_MESSAGE)


def _entry(record: "TechLeadRunRecord") -> TechLeadRunActivityEntry:
    return TechLeadRunActivityEntry(
        run_key=record.run_key,
        flavor_label=_FLAVOR_LABELS[record.flavor],
        phase=record.phase.value,
        phase_label=_PHASE_LABELS[record.phase],
        tone=_PHASE_TONES[record.phase],
        started_at=record.started_at.isoformat(),
        ended_at=record.ended_at.isoformat() if record.ended_at else "",
        subject_issue_number=record.subject_issue_number,
        subject_title=record.subject_title,
        detail=record.detail,
        findings=record.findings,
        proposals=record.proposals,
        run_id=record.run_id,
        session_name=record.session_name,
    )


__all__ = [
    "ACTIVITY_LIMIT",
    "EMPTY_MESSAGE",
    "TechLeadActivityView",
    "TechLeadRunActivityEntry",
    "read_tech_lead_activity",
]

File: src/issue_orchestrator/static/js/dashboard/tech_lead_activity.js
// Tech-lead activity panel — ADR-0033's local visibility surface (#6858).
//
// ``tech_lead_runs.js`` owns what the operator can DO; this owns what the tech
// lead has already DONE. It renders the typed ``techLeadActivity`` payload the
// dashboard view model publishes: one row per recorded run, with its scope, its
// phase, what it produced, and the subject it REFERENCED (never owned).
//
// Nothing is decided here. Phase words, tones, flavor names and the empty-state
// sentence all arrive from the server projection, so the browser cannot invent
// a second vocabulary for run state — the same rule the run-action affordances
// follow.
//
// Accessibility:
//
// * The panel is a native <details> disclosure with a real <summary>, so it is
//   keyboard reachable, has an accessible name, and reports expanded state
//   without any ARIA of our own.
// * Runs are a real <ul>/<li> list, so a screen reader announces how many there
//   are and where it is in them.
// * Phase is ALWAYS rendered as text (``phaseLabel``) next to its tone class —
//   colour is never the only signal.
// * Updates replace only the list's rows, so an operator's open panel and
//   keyboard focus survive live SSE-driven refreshes.

function techLeadActivityState() {
    return (window.dashboardData && window.dashboardData.techLeadActivity) || null;
}

// Pure: the runs to render, newest first (the server already ordered them).
function techLeadActivityEntries(activity) {
    return (activity && Array.isArray(activity.entries)) ? activity.entries : [];
}

// Pure: "#123 Some title", or '' when the run referenced no GitHub object.
// A whole-board review's subject is the board, and claiming an issue number it
// does not have would point the operator at the wrong thing.
function techLeadActivitySubjectText(entry) {
    const number = Number(entry.subjectIssueNumber) || 0;
    if (!number) {
        return '';
    }
    const title = entry.subjectTitle ? ` ${entry.subjectTitle}` : '';
    return `#${number}${title}`;
}

// Pure: "2 findings · 1 proposal", or '' when the run produced neither.
// Singular/plural is decided here rather than server-side because it is pure
// presentation of two numbers the projection already publishes.
function techLeadActivityProducedText(entry) {
    const parts = [];
    const findings = Number(entry.findings) || 0;
    const proposals = Number(entry.proposals) || 0;
    if (findings) {
        parts.push(`${findings} finding${findings === 1 ? '' : 's'}`);
    }
    if (proposals) {
        parts.push(`${proposals} proposal${proposals === 1 ? '' : 's'}`);
    }
    return parts.join(' · ');
}

// Pure: one <li> for a recorded run. Every interpolated value is escaped: the
// subject title and the decision summary are agent- and GitHub-authored text.
function renderTechLeadActivityRow(entry) {
    const cells = [
        `<span class="tla-flavor">${escapeHtml(entry.flavorLabel || '')}</span>`,
        `<span class="tla-phase tla-phase--${escapeAttr(entry.tone || 'muted')}">`
        + `${escapeHtml(entry.phaseLabel || '')}</span>`,
    ];
    const subject = techLeadActivitySubjectText(entry);
    if (subject) {
        cells.push(`<span class="tla-subject">${escapeHtml(subject)}</span>`);
    }
    const started = formatTimestamp(entry.startedAt, '');
    if (started) {
        cells.push(`<span class="tla-started">${escapeHtml(started)}</span>`);
    }
    const produced = techLeadActivityProducedText(entry);
    if (produced) {
        cells.push(`<span class="tla-produced">${escapeHtml(produced)}</span>`);
    }
    if (entry.detail) {
        cells.push(
            `<span class="tla-detail" title="${escapeAttr(entry.detail)}">`
            + `${escapeHtml(entry.detail)}</span>`,
        );
    }
    return `<li class="tla-row">${cells.join('')}</li>`;
}

// Pure: the whole list body — rows, or the server's empty-state sentence.
function techLeadActivityRowsHtml(activity) {
    const entries = techLeadActivityEntries(activity);
    if (!entries.length) {
        const message = (activity && activity.emptyMessage) || '';
        return `<li class="tla-row tla-row--empty">${escapeHtml(message)}</li>`;
    }
    return entries.map(renderTechLeadActivityRow).join('');
}

// Pure: the count shown beside the panel's name.
function techLeadActivityCountText(activity) {
    const count = techLeadActivityEntries(activity).length;
    return count ? String(count) : '';
}

// Surgically update the panel. The <details>/<summary> nodes are never
// replaced, so an operator's expanded panel and keyboard focus survive every
// live refresh; only the <ul> rows and the count change.
function updateTechLeadActivityPanel(activity) {
    const panel = document.getElementById('techLeadActivityPanel');
    // No payload yet is NOT an empty history: before the first view-model
    // arrives we have observed nothing, so the panel is left as rendered rather
    // than told there are no runs. Only a real projection writes rows.
    if (!panel || !activity) {
        return;
    }
    const list = document.getElementById('techLeadActivityList');
    if (list) {
        list.innerHTML = techLeadActivityRowsHtml(activity);
    }
    const count = document.getElementById('techLeadActivityCount');
    if (count) {
        count.textContent = techLeadActivityCountText(activity);
    }
}

// Convenience for the load/refresh hooks.
function renderTechLeadActivityFromDashboardData() {
    updateTechLeadActivityPanel(techLeadActivityState());
}
